### DISCUSSION TITLE
Script class needs to have a virtual new() method

### DISCUSSION THREAD

**Discussion by blackears** [godotengine/godot-proposals#6887] [Scripting]
Right now the Script class has no method for creating new instances of the script it points to. Both of the nodes that currently inherit from it - GDScript and CSharpScript - do and I would presume any other derived nodes you add in the future would too. While Godot's duck typing does allow me to just call .new() on the Script node, it would be nice to have this formally added to the Script class so there is no ambiguity about how to create new instances from a script without having to know precisely which type of script you are instancing.

**Comment by theraot** [comment]
It took me a minute to parse, so I'll reiterate it in case that helps.

`Script` does not have a `new` method. Instead it is part of of `GDScript` ([`new`](https://docs.godotengine.org/en/stable/classes/class_gdscript.html#class-gdscript-method-new)) and `CSharpScript` ([`new`](https://docs.godotengine.org/en/stable/classes/class_csharpscript.html#class-csharpscript-method-new)). As a result, to safely call them, we must know the type of `Script` we are dealing with, in particular if there are new types of scripts added.

Thus, adding `new` to `Script` and requiring future or third party language support to use it would make them easier to consume.
